### PR TITLE
MINOR: Code Cleanup

### PR DESCRIPTION
**Pull request by vahidhashemian**
Clean up includes:

- Switching try-catch-finally blocks to try-with-resources when possible
- Removing some seemingly unnecessary `@SuppressWarnings` annotations
- Resolving some Java warnings
- Closing unclosed Closable objects
- Removing unused code

### CODE REVIEW

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java [hunk 1]
```diff
@@ -85,16 +85,17 @@ public void testConstructorFailureCloseResource() {
 
         final int oldInitCount = MockMetricsReporter.INIT_COUNT.get();
         final int oldCloseCount = MockMetricsReporter.CLOSE_COUNT.get();
+        KafkaProducer<byte[], byte[]> producer = null;
         try {
```
Comment (baluchicken) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: Why shouldn't we use the try with resources java 7 feature here?

Comment (vahidhashemian) [reply clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: Thanks. That's a great suggestion. I'll update the patch to use it.

**Comment by vahidhashemian** [comment]
cc @ijuma for feedback. Thanks.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5360/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5345/
Test PASSed (JDK 8 and Scala 2.12).

File: clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java [hunk 2]
```diff
@@ -77,6 +76,7 @@ public void testPartitioner() throws Exception {
         assertEquals("Partition should be correct", 1, metadata.get().partition());
         producer.clear();
         assertEquals("Clear should erase our history", 0, producer.history().size());
+        producer.close();
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java:90]: It's odd that we close this producer, but not the one assigned to the field with the same name. Do we care about closing `MockProducer`? Did you use a tool to find cases where `Closeable`s are not being closed?

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/FileOffsetBackingStore.java [hunk 3]
```diff
@@ -72,8 +72,10 @@ private void load() {
         try {
             ObjectInputStream is = new ObjectInputStream(new FileInputStream(file));
```
Comment (ijuma) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/FileOffsetBackingStore.java]: Why not try with resources?

File: tools/src/main/java/org/apache/kafka/tools/ClientCompatibilityTest.java [hunk 4]
```diff
@@ -372,8 +372,10 @@ public void testConsume(final long prodTimeMs) throws Throwable {
             new ClientCompatibilityTestDeserializer(testConfig.expectClusterId);
         final KafkaConsumer<byte[], byte[]> consumer = new KafkaConsumer<>(consumerProps, deserializer, deserializer);
         final List<PartitionInfo> partitionInfos = consumer.partitionsFor(testConfig.topic);
-        if (partitionInfos.size() < 1)
+        if (partitionInfos.size() < 1) {
+            consumer.close();
```
Comment (ijuma) [inline tools/src/main/java/org/apache/kafka/tools/ClientCompatibilityTest.java]: Maybe easier to use try with resources?

**Review by ijuma — COMMENTED** [review]
Thanks for cleaning up the code. Left some feedback. We'd also want someone more familiar with Streams to take a look at the Streams part. cc @enothereska @guozhangwang

File: clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java (hunk 2 above)
Comment (vahidhashemian) [reply clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java:90]: Thanks for catching this. I think I missed some `close()` calls on the member variable. I basically looked for usages of `Closable` objects (did not use a tool, so I may have missed some). For this particular one, `MockProducer` implements the `Producer` (which is `Closable`). And it looks like I need to close both the local and the member variables. I'll add that in the next commit.

I think I'll add the `close()` for member variable in a junit `@After` method so the producer is also closed when we are expecting exceptions. That means in some methods that `close()` is called earlier for the test we need to define additional local variables.

I hope that makes sense. Please advise if you can think of a better approach. Thanks.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/FileOffsetBackingStore.java (hunk 3 above)
Comment (vahidhashemian) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/FileOffsetBackingStore.java]: Yes, I had missed this one. Thanks.

File: tools/src/main/java/org/apache/kafka/tools/ClientCompatibilityTest.java (hunk 4 above)
Comment (vahidhashemian) [reply tools/src/main/java/org/apache/kafka/tools/ClientCompatibilityTest.java]: This one seems a bit non-trivial to use try with resources. I'll give it a shot :)

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java [hunk 5]
```diff
@@ -223,7 +223,7 @@ public KafkaProducer(Properties properties, Serializer<K> keySerializer, Seriali
                 keySerializer, valueSerializer);
     }
 
-    @SuppressWarnings({"unchecked", "deprecation"})
+    @SuppressWarnings("unchecked")
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:305]: Hmm.. I'm wondering why it was added at the first place :P

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 6]
```diff
@@ -53,7 +53,6 @@
  * @see KGroupedStream
  * @see KStreamBuilder#stream(String...)
  */
-@SuppressWarnings("unused")
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:56]: This is used as some of the apis are never used within the library itself, curious to learn what's the motivation of removing it.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java [hunk 7]
```diff
@@ -58,7 +58,6 @@
     private final Map<TopicPartition, Long> checkpointedOffsets;
     private final Map<String, StateRestoreCallback> restoreCallbacks; // used for standby tasks, keyed by state topic name
     private final Map<String, String> storeToChangelogTopic;
-    private final boolean eosEnabled;
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java]: cc @mjsax

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfo.java [hunk 8]
```diff
@@ -128,6 +128,7 @@ public static AssignmentInfo decode(ByteBuffer data) {
             if (version < 0 || version > CURRENT_VERSION) {
                 TaskAssignmentException ex = new TaskAssignmentException("Unknown assignment data version: " + version);
                 log.error(ex.getMessage(), ex);
+                in.close();
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfo.java]: Should we also close the input stream at line 157 before returning as well?

**Review by guozhangwang — COMMENTED** [review]
lgtm. also cc'ing @mjsax for another look.

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java (hunk 5 above)
Comment (vahidhashemian) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:305]: It was added a long time ago by [this PR](https://github.com/apache/kafka/commit/241c3ebb2803f1e09306fb06f20a66e7a60ca3c8), due to deprecated producer configs, such as `ProducerConfig.BLOCK_ON_BUFFER_FULL_CONFIG`, that no longer exist.

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java (hunk 6 above)
Comment (vahidhashemian) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:56]: My IDE identified this line as unnecessary. I guess it doesn't check whether the APIs are used or not. I can put it back if you think it should be there.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfo.java (hunk 8 above)
Comment (vahidhashemian) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfo.java]: Yes, actually I'll try to use try with resources here in the next update. Thanks for catching it.

**Review by vahidhashemian — COMMENTED** [review]
@ijuma @guozhangwang Thanks for reviewing the PR. I tried to address your concerns in the new commit. Please also see my responses inline, and advise if you still see issues. Thanks.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5405/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5420/
Test PASSed (JDK 7 and Scala 2.11).

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java (hunk 7 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java]: Variable is unused so that's ok. Seems to be slipped to remove it after some code refactoring (cannot recall anymore why I introduced it originally).

**Comment by vahidhashemian** [comment]
@ijuma @guozhangwang @mjsax Thanks for your feedback on this PR. Do you see anything that still needs to be addressed? Thanks.

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java [hunk 9]
```diff
@@ -369,7 +368,7 @@ public void testHeaders() throws Exception {
         MemberModifier.field(KafkaProducer.class, "metadata").set(producer, metadata);
 
         String topic = "topic";
-        Collection<Node> nodes = Collections.singletonList(new Node(0, "host1", 1000));
+        Collections.singletonList(new Node(0, "host1", 1000));
```
Comment (mjsax) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: It seem this can be removed completely? Compare line 375 (I guess, `nodes` was use there at some point but not anymore)

File: clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java (hunk 2 above)
Comment (mjsax) [reply clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java:90]: Why don't we use try-with-resource pattern?

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java (hunk 9 above)
Comment (vahidhashemian) [reply clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: That's correct. I'll remove the line in the next update. Thanks.

File: clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java (hunk 2 above)
Comment (vahidhashemian) [reply clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java:90]: For this PR I tried to use try-with-resource pattern instead of an existing try-catch block, when possible. I'm not sure about the best-practice here, but my feeling is the code is more readable this way, plus I'm not sure if try-with-resource adds much value here. Please advise if you disagree.

**Review by vahidhashemian — COMMENTED** [review]
@mjsax Thanks for taking another look. Please see my responses inline. I'll send an update shortly to also resolve the conflict.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5674/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5660/
Test PASSed (JDK 8 and Scala 2.12).

File: clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java (hunk 2 above)
Comment (mjsax) [reply clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java:90]: Don't think readability would suffer. For `MockProducer` is might also not be bad atm if `close()` is not called in case the test fails with an exception and thus `close()` is never called. It just feel "right" to make sure to call `close()` for all classes that are `Closable` -- even in case of failure. This might also be more "future prove" in case `close()` implementation changes at some point. But I also don't insist on the change.

Comment (vahidhashemian) [reply clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java:90]: Sure, that sounds good. I think this pattern could be applied in other places (files) too, but I'll skip them for now. I'll submit a new commit. Please let me know if you still see room for improvement. Thanks.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5689/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5675/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by vahidhashemian** [comment]
Any other feedback on this PR?
cc @hachikuji

File: clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java [hunk 10]
```diff
@@ -51,9 +52,12 @@
     private final ProducerRecord<byte[], byte[]> record1 = new ProducerRecord<>(topic, "key1".getBytes(), "value1".getBytes());
     private final ProducerRecord<byte[], byte[]> record2 = new ProducerRecord<>(topic, "key2".getBytes(), "value2".getBytes());
 
+    @After
+    public void close() {
+        producer.close();
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java]: Since we seem to be replacing many of the global references below, do we still need this? Perhaps we should just replace them all? Or perhaps we should use the "registration" approach that we use in some of the scala tests in order to make cleanup consistent between all cases?

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/FileOffsetBackingStore.java [hunk 11]
```diff
@@ -69,19 +69,19 @@ public synchronized void stop() {
 
     @SuppressWarnings("unchecked")
     private void load() {
-        try {
-            ObjectInputStream is = new ObjectInputStream(new FileInputStream(file));
+        try (ObjectInputStream is = new ObjectInputStream(new FileInputStream(file))) {
             Object obj = is.readObject();
-            if (!(obj instanceof HashMap))
+            if (!(obj instanceof HashMap)) {
+                is.close();
```
Comment (hachikuji) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/FileOffsetBackingStore.java]: Why is this needed? Doesn't the `try` ensure that the input stream will be closed?

File: connect/transforms/src/test/java/org/apache/kafka/connect/transforms/CastTest.java [hunk 12]
```diff
@@ -74,6 +79,7 @@ public void castWholeRecordKeyWithSchema() {
 
         assertEquals(Schema.Type.INT8, transformed.keySchema().type());
         assertEquals((byte) 42, transformed.key());
+        xform.close();
```
Comment (hachikuji) [inline connect/transforms/src/test/java/org/apache/kafka/connect/transforms/CastTest.java]: Any reason we don't use the "try with resources" in these other cases as well? Maybe since the constructor has no arguments, we can just use a global instance and close in an `@After` method? Same for the other transforms below.

**Review by hachikuji — COMMENTED** [review]
Thanks for the patch. Left a few minor comments.

File: clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java (hunk 10 above)
Comment (vahidhashemian) [reply clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java]: I agree there is some inconsistency as a result of introducing `try-with-resources` blocks. Even though quite a few global references are replaced, there are still a lot of test methods that use global references. I could introduce `try-with-resources` to all methods to make it consistent and get rid of the `@After` method above. By "registration" did you mean something like producers and consumers setup and teardown in [`IntegrationTestHarness.scala`](https://github.com/apache/kafka/blob/0bede30ada6d5719950ad35a69c586f0ba9c4f7e/core/src/test/scala/integration/kafka/api/IntegrationTestHarness.scala)?

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/FileOffsetBackingStore.java (hunk 11 above)
Comment (vahidhashemian) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/FileOffsetBackingStore.java]: Yes, you are right, as `close()` is called even when throwing the exception. Thanks for catching it.

File: connect/transforms/src/test/java/org/apache/kafka/connect/transforms/CastTest.java (hunk 12 above)
Comment (vahidhashemian) [reply connect/transforms/src/test/java/org/apache/kafka/connect/transforms/CastTest.java]: Sure, I'll do that in the next patch, but I think in some of them we run into the same issue as the `MockProducerTest` issue you identified above. I hope I did not misunderstand your point.

**Review by vahidhashemian — COMMENTED** [review]
@hachikuji Thanks for your feedback. I'll try to submit an update to address two of your comments, while wait for your clarification on the third one.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6036/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6051/
Test PASSed (JDK 7 and Scala 2.11).

File: clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java (hunk 10 above)
Comment (hachikuji) [reply clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java]: Yes, that's what I was thinking. Then we can remove the global references and use the same pattern consistently.

File: connect/transforms/src/test/java/org/apache/kafka/connect/transforms/CastTest.java [hunk 13]
```diff
@@ -34,42 +35,54 @@
...
+    public void teardown() {
+        xformKey.close();
+        xformValue.close();
+    }
 
     @Test(expected = ConfigException.class)
     public void testConfigEmpty() {
-        final Cast<SourceRecord> xform = new Cast.Key<>();
-        xform.configure(Collections.singletonMap(Cast.SPEC_CONFIG, ""));
+        try (final Cast<SourceRecord> xform = new Cast.Key<>()) {
```
Comment (hachikuji) [inline connect/transforms/src/test/java/org/apache/kafka/connect/transforms/CastTest.java]: We could instead use `xformKey`? Same for others below.

File: clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java (hunk 10 above)
Comment (vahidhashemian) [reply clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java]: I submitted an update, in which I removed the global variable and used `try-with-resources` blocks when it was possible. The tests now seem to have a consistent pattern.  I hope I did not misunderstand your comment.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6087/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6072/
Test PASSed (JDK 8 and Scala 2.12).

File: clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java [hunk 14]
```diff
@@ -47,22 +47,21 @@
...
     public void testAutoCompleteMock() throws Exception {
-        Future<RecordMetadata> metadata = producer.send(record1);
-        assertTrue("Send should be immediately complete", metadata.isDone());
-        assertFalse("Send should be successful", isError(metadata));
-        assertEquals("Offset should be 0", 0L, metadata.get().offset());
-        assertEquals(topic, metadata.get().topic());
-        assertEquals("We should have the record in our history", singletonList(record1), producer.history());
-        producer.clear();
-        assertEquals("Clear should erase our history", 0, producer.history().size());
+        try (MockProducer<byte[], byte[]> producer = new MockProducer<>(true, new MockSerializer(), new MockSerializer())) {
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java]: Have to say it's a bit annoying to have all these diffs just to close the mock producer (which does not have any resources to close). I wonder if we can find a lighter pattern. What I was trying to suggest before was to continue to use the shared `producer` with an `@After`. We can use a factory method to create it in each method. For example:
```java
@After
public void cleanup() {
  if (this.producer != null) this.producer.close()
}

private void buildMockProducer(boolean autoComplete) {
  this.producer = new MockProducer<>(autocomplete, new MockSerializer(), new MockSerializer()))
}
```
We can add overloads to `buildMockProducer` as needed. What do you think?

Comment (vahidhashemian) [reply clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java]: Yup, this works too as another way of bringing consistency into these test. Sorry for misunderstanding your suggestion earlier. I'll submit an update shortly.

File: clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java [hunk 15]
```diff
@@ -70,185 +81,170 @@ public void testPartitioner() throws Exception {
...
         Cluster cluster = new Cluster(null, new ArrayList<Node>(0), asList(partitionInfo0, partitionInfo1),
                 Collections.<String>emptySet(), Collections.<String>emptySet());
-        try (MockProducer<String, String> producer = new MockProducer<>(cluster, true, new DefaultPartitioner(), new StringSerializer(), new StringSerializer())) {
-            ProducerRecord<String, String> record = new ProducerRecord<>(topic, "key", "value");
-            Future<RecordMetadata> metadata = producer.send(record);
-            assertEquals("Partition should be correct", 1, metadata.get().partition());
-            producer.clear();
-            assertEquals("Clear should erase our history", 0, producer.history().size());
-        }
+        MockProducer<String, String> producer = new MockProducer<>(cluster, true, new DefaultPartitioner(), new StringSerializer(), new StringSerializer());
```
Comment (vahidhashemian) [inline clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java]: @hachikuji The only anomaly is here, where the producer type is different. Not sure if there is a cleaner way to handle it.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6136/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6152/
Test FAILed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6137/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6153/
Test PASSed (JDK 7 and Scala 2.11).

**Review by hachikuji — APPROVED** [review]
LGTM
